File: JS/Tree/1609.Even Odd Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isEvenOddTree = function(root) {
    const stack = [root];
    let count = 1;
    let current = 0;
    let depth = 0;
    let previous = null;
    while(stack.length) {
        const node = stack.shift();
        if (depth % 2 === 0) {
            if (node.val % 2 === 0) {
                return false;
            }
            if (previous !== null && previous >= node.val) {
                return false;
            }
        } else {
            if (node.val %2 === 1) {
                return false;
            }
            if (previous !== null && previous <= node.val) {
                return false;
            }
        }
        current++;
        previous = node.val;
        if (node.left) {
            stack.push(node.left);
        }
        if (node.right) {
            stack.push(node.right);
        }
        if (current === count) {
            current = 0;
            depth++;
            count = stack.length;
            previous = null;
        }
    }
    return true;
};

File: Golang/Tree/1609.Even-odd-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isEvenOddTree(root *TreeNode) bool {
	if root.Val%2 == 0 {
		return false
	}
	level := 0
	stack := []*TreeNode{root}
	for len(stack) > 0 {
		children := []*TreeNode{}
		level++
		for _, s := range stack {
			if s.Left != nil {
				children = append(children, s.Left)
			}
			if s.Right != nil {
				children = append(children, s.Right)
			}
		}
		if len(children) > 0 {
			for i := 0; i < len(children); i++ {
				if level%2 == 0 && children[i].Val%2 == 0 {
					return false
				}
				if level%2 == 1 && children[i].Val%2 == 1 {
					return false
				}
				if i > 0 {
					if level%2 == 0 && children[i].Val <= children[i-1].Val {
						return false
					}
					if level%2 == 1 && children[i].Val >= children[i-1].Val {
						return false
					}
				}
			}
		}
		stack = children
	}
	return true
}
